mgr/cephadm: more rgw streamlining

Daniel-Pivonka opened the pull request:
IMPORTANT: some refactoring will be required when https://github.com/ceph/ceph/pull/38284 is merged.

**goal: make deploying a rgw service a one step process**

prior to this it was a multi step process to get everything setup

```
radosgw-admin realm create --rgw-realm=test_realm --default
radosgw-admin zonegroup create --rgw-zonegroup=default --rgw-realm=test_realm --master --default
radosgw-admin zone create --rgw-zonegroup=default --rgw-zone=test_zone --rgw-realm=test_realm --master --default
radosgw-admin period update --rgw-realm=test_realm --commit
ceph orch apply rgw test_realm test_zone 
radosgw-admin user create --uid=test_user --display-name=TEST_USER --system 
ceph dashboard set-rgw-api-access-key 
ceph dashboard set-rgw-api-secret-key
```

i have tried to automate this process as much as i can but the hard part is dealing with multiple realms and zones.

i believe this logic will prevent cephadm from messing with realms, zonegroups, and zone the the user has manually created
and will only create things when it's safe to do so. 

it will create a dashboard user if the keys arent set and a user doesnt exist.

this will streamline simple use cases (1 realm, 'default' zonegroup, multiple zone)
more complex use cases with multiple realms, zonegroups will need to be handled manually

this will also prevent invalid configs 
if zone_b exists in realm_b this would be raise an error ``ceph orch apply rgw realm_a zone_b``



_general info:_

realms contain zonegroups and zonegroups contain zones    realms(zonegroups(zones))

realms need a master zonegroup and zone, the first ones created are the masters usually

every realm, zonegroup, and zone have a name and these must be unique across the entire cluster (no 2 zones can have the same name for example)

commands like ``radosgw-admin zone list --rgw-zone=<zone>`` return all the zones in the cluster not just in a realm

to get zones just in a specific realms you need to get the period and extract it from the json it returns ``radosgw-admin period get --rgw-realm=test_realm`` 

the dashboard expects the zonegroup to be called 'default', bucket create requests contain a flag like this '--rgw-zonegroup=default' hard coded in 

dashboard only deals with one realm at a time. a user is connected to a realm

Related PRs:
#36162
#36496 

Fixes:
https://tracker.ceph.com/issues/43681
https://tracker.ceph.com/issues/44926
https://tracker.ceph.com/issues/44605

Signed-off-by: Daniel-Pivonka <dpivonka@redhat.com>

alfonsomthd commented:
@cbodley  Is it advisable to create the system user just after the master zone creation?

mattbenjamin commented:
@alimaredia :)

cbodley commented:
> @cbodley Is it advisable to create the system user just after the master zone creation?

what's the alternative? what do the multisite docs do?

alimaredia commented:
@alfonsomthd that is what we are doing in ceph-ansible: https://github.com/ceph/ceph-ansible/blob/master/roles/ceph-rgw/tasks/multisite/master.yml#L61

alfonsomthd commented:
> > @cbodley Is it advisable to create the system user just after the master zone creation?
> 
> what's the alternative? what do the multisite docs do?

@cbodley In the CLI example the user is created after updating the period and deploying the daemon.
My question was about creating the user before those steps.
I remember facing some problems in the past when creating several realms/zonegroups/zones etc and creating the user as a last step.

alfonsomthd commented:
> @alfonsomthd that is what we are doing in ceph-ansible: https://github.com/ceph/ceph-ansible/blob/master/roles/ceph-rgw/tasks/multisite/master.yml#L61

@alimaredia Thanks!  I'll take a look.

cbodley commented:
this pr seems to conflate the dashboard user with "the system user" in multisite docs. we should keep the two separate! the dashboard user doesn't need to be a `--system` user; it could be `--admin` (which, like system, has global read/write permissions), or ideally even just a normal user with caps[1] for the specific admin APIs it uses

[1] https://docs.ceph.com/en/latest/radosgw/admin/#add-remove-admin-capabilities

it looks like the goal of this PR is just to create a single zone in isolation, so we shouldn't need to bother with the system user for multisite at all. the dashboard user can be created just after committing the period

theanalyst commented:
Additionally at this point I'd suggest adding a rgw_zonegroup in the yaml and args and allowing a creation of that, since we're hardcoding the zonegroup default everywhere, and that'd have consequences if you're planning to create a non "default" named realm/zone/zonegroup as the system would end up creating a new zone with the spec. under the default zonegroup and not the user specified one!

sebastian-philipp commented:
> adding a rgw_zonegroup in the yaml and args and allowing a creation of that, since we're hardcoding the zonegroup default everywhere, and that'd have consequences if you're planning to create a non "default" named realm/zone/zonegroup as the system would end up creating a new zone with the spec. under the default zonegroup and not the user specified one!

https://tracker.ceph.com/issues/48340

sebastian-philipp commented:
ping @Daniel-Pivonka

sebastian-philipp commented:
jenkins test make check

Daniel-Pivonka commented:
finally getting back to this! added rgw zonegroup arg to apply rgw and switched dashboard user from 'system' to 'admin'. need to rebase and make changes for #38284

sebastian-philipp commented:
make check:

```
cephadm/module.py:30: note: In module imported here:
cephadm/services/cephadmservice.py: note: In member "config" of class "RgwService":
cephadm/services/cephadmservice.py:542: error: Function is missing a return type annotation
cephadm/services/cephadmservice.py: note: In member "get_client_keyring" of class "RgwService":
cephadm/services/cephadmservice.py:626: error: Function is missing a return type annotation
cephadm/services/cephadmservice.py: note: In member "create_realm_zonegroup_zone" of class "RgwService":
cephadm/services/cephadmservice.py:633: error: Function is missing a return type annotation
cephadm/services/cephadmservice.py: note: In function "create_realm_zonegroup_zone":
cephadm/services/cephadmservice.py:659: error: Function is missing a return type annotation
cephadm/services/cephadmservice.py: note: In function "get_period":
cephadm/services/cephadmservice.py:659: error: Function is missing a return type annotation
cephadm/services/cephadmservice.py: note: In member "create_realm_zonegroup_zone" of class "RgwService":
cephadm/services/cephadmservice.py:705: error: Name 'create_realm' already defined on line 650
cephadm/services/cephadmservice.py: note: In function "create_realm_zonegroup_zone":
cephadm/services/cephadmservice.py:705: error: Function is missing a return type annotation
cephadm/services/cephadmservice.py:705: note: Use "-> None" if function does not return a value
cephadm/services/cephadmservice.py: note: In function "create_realm":
cephadm/services/cephadmservice.py:705: error: Function is missing a return type annotation
cephadm/services/cephadmservice.py:705: note: Use "-> None" if function does not return a value
cephadm/services/cephadmservice.py: note: In function "create_realm_zonegroup_zone":
cephadm/services/cephadmservice.py:714: error: Function is missing a return type annotation
cephadm/services/cephadmservice.py:714: note: Use "-> None" if function does not return a value
cephadm/services/cephadmservice.py: note: In function "create_zonegroup":
cephadm/services/cephadmservice.py:714: error: Function is missing a return type annotation
cephadm/services/cephadmservice.py:714: note: Use "-> None" if function does not return a value
cephadm/services/cephadmservice.py: note: In member "create_realm_zonegroup_zone" of class "RgwService":
cephadm/services/cephadmservice.py:775: error: Argument 2 to "OrchestratorError" has incompatible type "Optional[str]"; expected "int"
cephadm/services/cephadmservice.py: note: In member "create_dashboard_user" of class "RgwService":
cephadm/services/cephadmservice.py:831: error: Function is missing a return type annotation
cephadm/services/cephadmservice.py: note: In function "create_dashboard_user":
cephadm/services/cephadmservice.py:833: error: Function is missing a return type annotation
cephadm/services/cephadmservice.py:833: note: Use "-> None" if function does not return a value
cephadm/services/cephadmservice.py: note: In function "create_user":
cephadm/services/cephadmservice.py:833: error: Function is missing a return type annotation
cephadm/services/cephadmservice.py:833: note: Use "-> None" if function does not return a value
cephadm/module.py: note: In member "_apply_service" of class "CephadmOrchestrator":
cephadm/module.py:2024: error: Unexpected keyword argument "get_hosts_func" for "HostAssignment"
cephadm/module.py:2026: error: "CephadmOrchestrator" has no attribute "_get_hosts"; maybe "get_hosts"?
cephadm/module.py:2031: error: Name 'HostPlacementSpec' is not defined
cephadm/module.py:2044: error: Name 'HostPlacementSpec' is not defined
cephadm/schedule.py:56: note: "HostAssignment" defined here
```

sebastian-philipp commented:
jenkins test make check

Daniel-Pivonka commented:
added doc changes. moving this from a draft to real PR

alfonsomthd commented on src/pybind/mgr/cephadm/services/cephadmservice.py:
```diff
@@ -699,41 +691,212 @@ def get_zones() -> List[str]:
+            'prefix': 'dashboard get-rgw-api-secret-key',
+        })
+
+        if not access_key and not secret_key:
```
```suggestion
        if not access_key or not secret_key:
```

alfonsomthd commented:
@Daniel-Pivonka Do you know if there is any integration test that checks the happy path (realm/zonegroup/zone correct creation & dashboard setup)?
If not, it would be great to add it to avoid regressions.

sebastian-philipp commented:
@Daniel-Pivonka where can I help to unstuck this?

Daniel-Pivonka commented:
rebased. 
added a qa test but not sure how to test it. any pointers on that would be helpful. 
facing this error when updating period ```2021-02-09T22:15:53.681+0000 7fdf3698c1c0  0 rgw_init_ioctx ERROR: librados::Rados::pool_create returned (34) Numerical result out of range (this can be due to a pool or placement group misconfiguration, e.g. pg_num < pgp_num or mon_max_pg_per_osd exceeded)
couldn't init storage provider```  investigating that currently

sebastian-philipp commented:
```
./tasks/cephadm_cases/test_rgw_autocreation.py:17:73: F821 undefined name 'BytesIO'
./tasks/cephadm_cases/test_rgw_autocreation.py:17:91: F821 undefined name 'BytesIO'
./tasks/cephadm_cases/test_rgw_autocreation.py:33:13: F821 undefined name 'time'
./tasks/cephadm_cases/test_rgw_autocreation.py:34:16: F821 undefined name 'expected_status'
./tasks/cephadm_cases/test_rgw_autocreation.py:37:19: F821 undefined name 'fail_msg'
./tasks/cephadm_cases/test_rgw_autocreation.py:58:13: F821 undefined name 'error'
./tasks/cephadm_cases/test_rgw_autocreation.py:62:13: F821 undefined name 'error'
```

```
cephadm/module.py:30: note: In module imported here,
cephadm/__init__.py:1: note: ... from here:
cephadm/services/cephadmservice.py: note: In member "config" of class "RgwService":
cephadm/services/cephadmservice.py:626: error: Argument 1 of "config" is incompatible with supertype "CephadmService"; supertype defines the argument type as "ServiceSpec"
cephadm/services/cephadmservice.py:626: note: This violates the Liskov substitution principle
cephadm/services/cephadmservice.py:626: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
cephadm/services/cephadmservice.py: note: In function "create_realm_zonegroup_zone":
cephadm/services/cephadmservice.py:771: error: Name 'rgw_id' is not defined
cephadm/services/cephadmservice.py: note: In member "create_realm_zonegroup_zone" of class "RgwService":
cephadm/services/cephadmservice.py:813: error: Name 'create_zonegroup' already defined on line 768
cephadm/services/cephadmservice.py: note: In function "create_user":
cephadm/services/cephadmservice.py:961: error: Incompatible types in assignment (expression has type "str", variable has type "bytes")
cephadm/services/cephadmservice.py:966: error: Incompatible types in assignment (expression has type "str", variable has type "bytes")
Found 5 errors in 1 file (checked 23 source files)
ERROR: InvocationError for command /home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/.tox/mypy/bin/mypy --config-file=../../mypy.ini -m alerts -m balancer -m cephadm -m crash -m dashboard -m devicehealth -m hello -m iostat -m mds_autoscaler -m mgr_module -m mgr_util -m mirroring -m orchestrator -m progress -m prometheus -m rook -m snap_schedule -m stats -m status -m telemetry -m test_orchestrator -m volumes -m zabbix (exited with code 1)
flake8 create: /home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/.tox/flake8
flake8 installdeps: flake8
flake8 installed: DEPRECATION: --find-links option in pip freeze is deprecated. pip 21.2 will remove support for this functionality. You can find discussion regarding this at https://github.com/pypa/pip/issues/9069.,-f /home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/wheelhouse,flake8==3.8.4,importlib-metadata==3.4.0,mccabe==0.6.1,pycodestyle==2.6.0,pyflakes==2.2.0,typing-extensions==3.7.4.3,zipp==3.4.0
flake8 run-test-pre: PYTHONHASHSEED='3852976376'
flake8 run-test: commands[0] | flake8 --config=tox.ini alerts balancer cephadm crash devicehealth hello orchestrator prometheus
cephadm/services/cephadmservice.py:750:13: F841 local variable 'e' is assigned to but never used
cephadm/services/cephadmservice.py:771:41: F821 undefined name 'rgw_id'
cephadm/services/cephadmservice.py:798:13: F841 local variable 'e' is assigned to but never used
cephadm/services/cephadmservice.py:813:9: F811 redefinition of unused 'create_zonegroup' from line 768
cephadm/services/cephadmservice.py:821:13: F841 local variable 'result' is assigned to but never used
cephadm/services/cephadmservice.py:973:13: F841 local variable 'e' is assigned to but never used
cephadm/services/cephadmservice.py:1000:13: F841 local variable 'e' is assigned to but never used
1     F811 redefinition of unused 'create_zonegroup' from line 768
1     F821 undefined name 'rgw_id'
```

alfonsomthd commented:
@Daniel-Pivonka Maybe you need to make some adaptations in order to set up properly the access/secret keys in dashboard settings for several daemons:
https://github.com/ceph/ceph/pull/38170

Daniel-Pivonka commented:
closing because #39877
